File: Linked List, Hash/leetcode/middle_of_the_linkedlist/876_middle_of_the_linkedlist.c
#include <stdio.h>
#include <stdlib.h>

// 노드 구조체 정의
struct ListNode {
    int val;
    struct ListNode* next;
};

// 중간 노드 찾기 함수
struct ListNode* middleNode(struct ListNode* head) {
    struct ListNode* slow = head;
    struct ListNode* fast = head;
    
    // fast가 끝에 도달할 때까지 반복
    while (fast != NULL && fast->next != NULL) {
        slow = slow->next;
        fast = fast->next->next;
    }
    
    // slow 포인터가 중간 노드를 가리킴
    return slow;
}

// 새로운 노드를 생성하는 함수
struct ListNode* createNode(int val) {
    struct ListNode* newNode = (struct ListNode*)malloc(sizeof(struct ListNode));
    newNode->val = val;
    newNode->next = NULL;
    return newNode;
}

// 리스트 출력 함수 (중간 노드부터 끝까지)
void printList(struct ListNode* node) {
    while (node != NULL) {
        printf("%d ", node->val);
        node = node->next;
    }
    printf("\n");
}

int main() {
    // 연결 리스트 생성 예시: 1 -> 2 -> 3 -> 4 -> 5
    struct ListNode* head = createNode(1);
    head->next = createNode(2);
    head->next->next = createNode(3);
    head->next->next->next = createNode(4);
    head->next->next->next->next = createNode(5);

    struct ListNode* middle = middleNode(head);

    // 중간 노드부터 출력
    printList(middle);  // 예상 출력: 3 4 5

    // 메모리 해제
    struct ListNode* temp;
    while (head != NULL) {
        temp = head;
        head = head->next;
        free(temp);
    }

    return 0;
}

File: Linked List, Hash/leetcode/middle_of_the_linkedlist/876_middle_of_the_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# 문제 설명 : singly linked list의 head가 input으로 주어질 때 중간노드 return 하기.

class Solution:
    # 첫번째 나이브한 풀이
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #length 구한 뒤 length//2로 중간노드 구함
        cnt=0
        cur=head
        while cur:
            cnt+=1
            cur=cur.next
        for i in range(cnt//2):
            head=head.next
        return head
    # 두번째 풀이
    def middleNode2(self, head: Optional[ListNode]) -> Optional[ListNode]:       
        # 투포인터
        '''
        두 포인터 사용: 빠른 포인터(fast)와 느린 포인터(slow)를 사용합니다.
				fast 포인터는 한 번에 두 칸씩 이동하고, slow 포인터는 한 번에 한 칸씩 이동합니다.
				반복: fast 포인터가 리스트의 끝에 도달하면 slow 포인터는 중간 지점에 도달하게 됩니다.
				리스트의 길이가 홀수이면 정확히 중간 노드를 가리키고, 짝수이면 두 번째 중간 노드를 가리키게 됩니다.
				'''
        fast=head
        slow=fast
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next
        return slow
       